### DISCUSSION TITLE
How to do client assigned IDs with Transaction builder?

### DISCUSSION THREAD

**Discussion by johnkwaters** [FirelyTeam/firely-net-sdk#2099] [Q&A]
I can't figure the correct syntax with TransactionBuilder to do client assigned IDs.
The server is a Smile CDR, and the developer says i need to create a Transaction bundle containing a patient resource with my desired ID.
They gave me an example (this is creating a resource of type Organization, not Patient, but same idea). Note that the HTTP is a POST, but the inner resource has a method of PUT...

![image](https://user-images.githubusercontent.com/3422637/170153885-a9cb6715-0a5e-477f-af93-aedb87b8f788.png)

However, the request generated by my code does a PUT on a Patient, rather than a POST on a Bundle...

How do I get the code to do what they want? Do I ignore the transaction builder and instead create a manual bundle with an entry and POST that? Seems the transationbuilder should be helping me do that if I can just understand how!

```
PUT https://fhir-st-mpi-clinical-dev.apps.okd.healthconcourse.com/Patient?_id=00000003
{
  "resourceType": "Patient",
  "id": "00000003",
  "identifier": [
    {
      "system": "http://ger911.com/fhir/hc",
      "value": "ORG-GER-00000003"
    }
  ],
  "active": true,
  "name": [
    {
      "family": "John",
      "given": [
        "Waters Test 2"
      ]
    }
  ]
}

```

**Comment by johnkwaters** [comment]
Update... this code generates something much closer (p is a new Patient Resource):

```
            var tx = new TransactionBuilder(fc.Endpoint, Bundle.BundleType.Transaction)
                .Update(p.Id, p);
            var b = tx.ToBundle();
            var res = await fc.CreateAsync(b);
```

This creates:

```
POST https://fhir-st-mpi-clinical-dev.apps.okd.healthconcourse.com/Bundle
{
  "resourceType": "Bundle",
  "type": "transaction",
  "entry": [
    {
      "resource": {
        "resourceType": "Patient",
        "id": "00000005",
        "identifier": [
          {
            "system": "http://ger911.com/fhir/hc",
            "value": "ORG-GER-00000005"
          }
        ],
        "active": true,
        "name": [
          {
            "family": "Waters",
            "given": [
              "005"
            ]
          }
        ]
      },
      "request": {
        "method": "PUT",
        "url": "Patient/00000005"
      }
    }
  ]
}

```

The only diff here is my POST ends in /Bundle, whereas the example I was given doesn't.

I get a Forbidden code back posting this.

**Comment by johnkwaters** [comment]
Seems there might be a server side setting for Client ID mode

![image](https://user-images.githubusercontent.com/3422637/170159811-1ba10f42-ea6a-4bc4-9bb5-4c03851af95e.png)

**Reply by johnkwaters** [reply]
Apparently this setting IS on... the vendor says that a POST of a Bundle should have the URL go to the root of the FHIR server, not have /Bundle on the end of it?

**Comment by johnkwaters** [comment]
Another update ; the issue #2100 resolved how to accomplish this POST:

```
            var tx = new TransactionBuilder(fc.Endpoint, Bundle.BundleType.Transaction)
                .Update(p.Id, p);
            var b = tx.ToBundle();

            var res = await fc.TransactionAsync(b);

```

But it returned a 500 error in my case.
I am still looking for guidance on how to do Client Assigned IDs the correct way.

**Reply by brianpos** [reply]
This is now sounding like an issue with the processing of this specific request and server handling.
Without the rest of the response, or trying it out on another server hard to tell.
You could try with the public Vonk?

**Reply by johnkwaters** [reply]
What is the public Vonk?
